Make ActiveRecord structure load/dump configurable

kirs opened the pull request:
Right now it's not possible to pass extra flags to
`mysqldump/pg_dump` when running `rake db:structure:dump` or `rake db:structure:load`

The following config variables (`structure_load_flags` and `structure_dump_flags`) make it better configurable.

Background: today at Shopify we have to monkey patch `ActiveRecord::Tasks::MySQLDatabaseTasks` to accommodate extra flags:

```ruby
module ActiveRecordDbCommandPatch
  def run_cmd(cmd, args, action)
    if cmd == "mysqldump"
      args.insert(0, "--no-defaults")
      args.concat(["--skip-add-drop-table"])
    elsif cmd == "mysql"
      args.insert(0, "--no-defaults")
    end

    super(cmd, args, action)
  end
end

ActiveRecord::Tasks::MySQLDatabaseTasks.prepend(ActiveRecordDbCommandPatch)
```

With this PR we can reduce amount of monkeypatches required to run Rails with MySQL at scale and make it more configurable.

rafaelfranca commented on activerecord/lib/active_record/core.rb:
```diff
@@ -134,6 +134,16 @@ def self.error_on_ignored_order_or_limit=(value)
+      ##
+      # :singleton-method:
+      # Extra flags passed to database CLI tool (mysqldump/pg_dump) when calling db:structure:dump
+      mattr_accessor :structure_dump_flags, instance_accessor: false
```
I don't think this configuration should be at `ActiveRecord::Base` Maybe it is a better fit for the `DatabaseTasks` class?

kirs replied:
Yeah, I wasn't sure about `AR::Base` either. Should it be configurable with:

```ruby
ActiveRecord::Tasks::DatabaseTasks.structure_dump_flags = ["--noop"]
ActiveRecord::Tasks::DatabaseTasks.structure_load_flags = ["--noop"]
```

?

rafaelfranca replied:
Yeah. That looks better for me.

kirs replied:
I'm a bit concerned about introducing extra place with a config, but I like it too.

kaspth commented on activerecord/lib/active_record/tasks/mysql_database_tasks.rb:
```diff
@@ -53,21 +53,23 @@ def collation
         args = prepare_command_options
         args.concat(["--execute", %{SET FOREIGN_KEY_CHECKS = 0; SOURCE #{filename}; SET FOREIGN_KEY_CHECKS = 1}])
         args.concat(["--database", "#{configuration['database']}"])
+        args.concat(Array(extra_flags)) if extra_flags.any?
```
This should probably just be `if extra_flags` like above.

kaspth commented on activerecord/test/cases/tasks/mysql_rake_test.rb:
```diff
@@ -323,6 +332,16 @@ def test_structure_dump_with_ssl
       end
+
+      private
+
```
✂️

kaspth commented on activerecord/test/cases/tasks/mysql_rake_test.rb:
```diff
@@ -340,6 +359,26 @@ def test_structure_load
+      end
+
+      private
+
```
✂️

kaspth commented on activerecord/test/cases/tasks/postgresql_rake_test.rb:
```diff
@@ -293,12 +309,31 @@ def test_structure_load
       end
+
+      private
+
```
✂️

kaspth commented on activerecord/test/cases/tasks/mysql_rake_test.rb:
```diff
@@ -340,6 +359,26 @@ def test_structure_load
+      def test_structure_load
+        filename = "awesome-file.sql"
+        Kernel.expects(:system).with("mysql", "--execute", %{SET FOREIGN_KEY_CHECKS = 0; SOURCE #{filename}; SET FOREIGN_KEY_CHECKS = 1}, "--database", "test-db", "--noop")
+          .returns(true)
```
We've been trying to move away from Mocha (admittedly stalled at the moment) and have introduced these assertions to help with that: https://github.com/rails/rails/blob/2fd3034bf1862bb8e40c0a0a162316b8ff564d0e/activesupport/lib/active_support/testing/method_call_assertions.rb

Can we switch to those in these new tests?

kirs committed:
Make ActiveRecord structure load/dump configurable

Without this patch it's impossible to pass extra flags to
mysqldump/pg_dump when running `rake db:structure:dump` or `load`

The following config variables (`structure_load_flags` and `structure_dump_flags`)
make it better configurable.

kirs commented:
Thanks everyone for review! I updated the PR.

kaspth commented:
Thanks!

stoivo commented:
I know it might be stupid to ask, but do you use this?

kirs commented:
Yep, we use `ActiveRecord::Tasks::DatabaseTasks.structure_dump_flags = ['--no-defaults', '--skip-add-drop-table']` at Shopify.

mecampbellsoup commented:
> Yep, we use `ActiveRecord::Tasks::DatabaseTasks.structure_dump_flags = ['--no-defaults', '--skip-add-drop-table']` at Shopify.

@kirs you just set this configuration in an initializer or something? Glad to have found this PR!
